odl/space/npy_tensors.py

```diff
@@ -0,0 +1,2336 @@
+
+        - The constant must be positive, otherwise it does not define an
+          inner product or norm, respectively.
+        """
+        super().__init__(const, impl='numpy', exponent=exponent,
+                         dist_using_inner=dist_using_inner)
+
+    def inner(self, x1, x2):
+        """Return the weighted inner product of ``x1`` and ``x2``.
+
+        Parameters
+        ----------
+        x1, x2 : `NumpyTensor`
+            Tensors whose inner product is calculated.
+
+        Returns
+        -------
+        inner : float or complex
+            The inner product of the two provided tensors.
+        """
+        if self.exponent != 2.0:
+            raise NotImplementedError('no inner product defined for '
+                                      'exponent != 2 (got {})'
+                                      ''.format(self.exponent))
+        else:
+            inner = self.const * _inner_default(x1, x2)
+            return x1.space.field.element(inner)
+
+    def norm(self, x):
+        """Return the weighted norm of ``x``.
+
+        Parameters
+        ----------
+        x1 : `NumpyTensor`
+            Tensor whose norm is calculated.
+
+        Returns
+        -------
+        norm : float
+            The norm of the tensor.
+        """
+        if self.exponent == 2.0:
+            return float(np.sqrt(self.const) * _norm_default(x))
+        elif self.exponent == float('inf'):
+            return float(self.const * _pnorm_default(x, self.exponent))
+        else:
+            return float((self.const ** (1 / self.exponent) *
+                          _pnorm_default(x, self.exponent)))
+
+    def dist(self, x1, x2):
+        """Return the weighted distance between ``x1`` and ``x2``.
+
+        Parameters
+        ----------
+        x1, x2 : `NumpyTensor`
+            Tensors whose mutual distance is calculated.
+
+        Returns
+        -------
+        dist : float
+            The distance between the tensors.
+        """
+        if self.dist_using_inner:
+            dist_squared = (_norm_default(x1) ** 2 + _norm_default(x2) ** 2 -
+                            2 * _inner_default(x1, x2).real)
+            if dist_squared < 0.0:  # Compensate for numerical error
+                dist_squared = 0.0
+            return float(np.sqrt(self.const * dist_squared))
+        elif self.exponent == 2.0:
+            return float(np.sqrt(self.const) * _norm_default(x1 - x2))
+        elif self.exponent == float('inf'):
+            return float(self.const * _pnorm_default(x1 - x2, self.exponent))
+        else:
+            return float((self.const ** (1 / self.exponent) *
+                          _pnorm_default(x1 - x2, self.exponent)))
+
+
+class NumpyTensorSpaceNoWeighting(NoWeighting,
+                                  NumpyTensorSpaceConstWeighting):
+
+    """Weighting of a `NumpyTensorSpace` with constant 1."""
+
+    # Implement singleton pattern for efficiency in the default case
+    __instanc
```

Comment:
The comment is w.r.t `MatrixOperator` where you assume the matrix is 2 or higher dimensional. I can easily see this working with 1 or even 0 dimensional arrays, so that condition is arbitrary.

The line "You use a 3 dimensional array later." refers to a comment that `MatrixOperator` input should be 2d, but you later use 3d. Also you later say that `axis` needs to be 1 or more, but the default is `0`.